src/realm/sync/noinst/client_impl_base.cpp

```diff
@@ -1402,26 +1402,30 @@ void Session::integrate_changesets(ClientReplication& repl, const SyncProgress&
 }
 
 
-void Session::on_integration_failure(const IntegrationException& error, DownloadBatchState batch_state)
+void Session::on_integration_failure(const IntegrationException& error)
 {
     REALM_ASSERT(m_state == Active);
-    if (batch_state == DownloadBatchState::LastInBatch) {
-        m_progress.download = m_download_progress;
-    }
     logger.error("Failed to integrate downloaded changesets: %1", error.what());
     m_conn.close_due_to_protocol_error(error.code());
 }
 
-void Session::on_changesets_integrated(version_type client_version, DownloadCursor download_progress,
-                                       DownloadBatchState batch_state)
+void Session::on_changesets_integrated(version_type client_version, const SyncProgress& progress)
 {
     REALM_ASSERT(m_state == Active);
-    REALM_ASSERT(download_progress.server_version >= m_download_progress.server_version);
-    m_download_batch_state = batch_state;
-    if (m_download_batch_state != DownloadBatchState::LastInBatch) {
-        return;
+    REALM_ASSERT(progress.download.server_version >= m_download_progress.server_version);
+    m_download_progress = progress.download;
+    bool upload_progressed = (progress.upload.client_version > m_progress.upload.client_version);
```

Comment:
This is what `update_progress()` used to do - but now we only do it when we successfully integrate changesets.